compiler-rt/lib/profile/InstrProfilingPlatformWindows.c

```diff
@@ -74,7 +74,14 @@ ValueProfNode *__llvm_profile_end_vnodes(void) { return &VNodesEnd; }
 ValueProfNode *CurrentVNode = &VNodesStart + 1;
 ValueProfNode *EndVNode = &VNodesEnd;
 
+uint64_t BuildIdLen = 16;
+COMPILER_RT_WEAK extern uint8_t __buildid[16];
 COMPILER_RT_VISIBILITY int __llvm_write_binary_ids(ProfDataWriter *Writer) {
+  if (*__buildid) {
```

Comment:
Shouldn't this be just `if (__buildid)`? Since the `__buildid` symbol is `extern` weak, it's going to be `NULL` if not defined elsewhere and we don't want to dereference `NULL` (I'm a bit surprised this wasn't actually caught by the `NO-BINARY-ID` test case).

Reply:
`__buildid` is an array. So, if it's not defined, it will be 16 byte array of 0. I'm basically checking the `__buildid[0]`.

Reply:
Not if it's an `extern` declaration, in that case it'll be just a pointer which may be `NULL` since the symbol is weak. Try the following example:
```c
#include <stdint.h>
__attribute__((weak)) extern uint8_t buildid[16];
int main() {
  if (*buildid)
    return 0;
  return 1;
}
```
```sh
$ clang a.c -o a.out
$ ./a.out
segmentation fault  ./a.out
```

Reply:
Oh I haven't realized weak symbols behave differently on Windows, ignore this comment and thank you the explanation.